src/Microsoft.ML.StandardLearners/Standard/SdcaBinary.cs

```diff
@@ -1817,75 +1805,62 @@ internal static class Defaults
         /// </summary>
         /// <param name="env">The environment to use.</param>
         /// <param name="options">Advanced arguments to the algorithm.</param>
-        internal StochasticGradientDescentClassificationTrainer(IHostEnvironment env, Options options)
+        /// <param name="loss">Loss function would be minimized.</param>
+        /// <param name="doCalibration">Set to true if a calibration step should be happen after training. Use false otherwise.</param>
+        internal SgdBinaryTrainerBase(IHostEnvironment env, OptionsBase options, IClassificationLoss loss = null, bool doCalibration = false)
             : base(env, options.FeatureColumn, TrainerUtils.MakeBoolScalarLabel(options.LabelColumn), options.WeightColumn.IsExplicit ? options.WeightColumn.Value : null)
         {
             options.Check(env);
-            _loss = options.LossFunction.CreateComponent(env);
-            Info = new TrainerInfo(calibration: !(_loss is LogLoss), supportIncrementalTrain: true);
+            Loss = loss;
+            Info = new TrainerInfo(calibration: doCalibration, supportIncrementalTrain: true);
             NeedShuffle = options.Shuffle;
             _options = options;
         }
 
-        protected override SchemaShape.Column[] GetOutputColumnsCore(SchemaShape inputSchema)
-        {
-            return new[]
-            {
-                new SchemaShape.Column(DefaultColumnNames.Score, SchemaShape.Column.VectorKind.Scalar, NumberDataViewType.Single, false, new SchemaShape(MetadataUtils.GetTrainerOutputMetadata())),
-                new SchemaShape.Column(DefaultColumnNames.Probability, SchemaShape.Column.VectorKind.Scalar, NumberDataViewType.Single, false, new SchemaShape(MetadataUtils.GetTrainerOutputMetadata(true))),
-                new SchemaShape.Column(DefaultColumnNames.PredictedLabel, SchemaShape.Column.VectorKind.Scalar, BooleanDataViewType.Instance, false, new SchemaShape(MetadataUtils.GetTrainerOutputMetadata()))
-            };
-        }
-
-        protected override BinaryPredictionTransformer<TScalarPredictor> MakeTransformer(TScalarPredictor model, DataViewSchema trainSchema)
-            => new BinaryPredictionTransformer<TScalarPredictor>(Host, model, trainSchema, FeatureColumn.Name);
+        protected override BinaryPredictionTransformer<TModelParameters> MakeTransformer(TModelParameters model, DataViewSchema trainSchema)
+            => new BinaryPredictionTransformer<TModelParameters>(Host, model, trainSchema, FeatureColumn.Name);
 
         /// <summary>
         /// Continues the training of a <see cref="SdcaBinaryTrainer"/> using an initial predictor and returns a <see cref="BinaryPredictionTransformer"/>.
         /// </summary>
-        public BinaryPredictionTransformer<TScalarPredictor> Fit(IDataView trainData, IPredictor initialPredictor)
+        public BinaryPredictionTransformer<TModelParameters> Fit(IDataView trainData, IPredictor initialPredictor = null)
```

Comment:
>= null [](start = 114, length = 6)

don't change it please.

Reply:
TrainTransformer is typed to TModelParameters, so we need this change.

---
In reply to: [257348893](https://github.com/dotnet/machinelearning/pull/2564#discussion_r257348893) [](ancestors = 257348893)

Reply:
If you look on my selection I was talking about `IPredictor initialPredictor = null` `=null` part

---
In reply to: [257351775](https://github.com/dotnet/machinelearning/pull/2564#discussion_r257351775) [](ancestors = 257351775,257348893)

Reply:
Ah. My bad.

---
In reply to: [257352673](https://github.com/dotnet/machinelearning/pull/2564#discussion_r257352673) [](ancestors = 257352673,257351775,257348893)